R/layout.R

```diff
@@ -108,3 +112,312 @@ layout_column_wrap <- function(
+      } else {
+        n_cols
+      }
+
+    if (is.null(bk_cols)) {
+      if (any(is.na(bk$width))) {
+        # we can't infer the number of columns from the widths, so assume 12
+        bk_cols <- 12
+      } else {
+        bk_cols <- sum(bk$width + bk$before + bk$after)
+      }
+    }
+
+    if (length(bk$width) > n_kids) {
+      # TODO: more informative warning
+      rlang::warn("Too many `widths` provided; truncating")
+    }
+
+    widths <- rep_len(bk$width, n_kids)
+    before <- rep_len(bk$before, n_kids)
+    after <- rep_len(bk$after, n_kids)
+
+    # Fill NA widths with equal shares of remaining space
+    idx_na <- which(is.na(widths))
+    if (length(idx_na) > 0) {
+      n_accounted <- sum(widths, na.rm = TRUE) + sum(before) + sum(after)
+      n_remaining <- bk_cols - n_accounted
+      widths[idx_na] <- max(1, floor(n_remaining / length(idx_na)))
+    }
+
+
+    cursor <- 0L
+    update_cursor <- function(incr, can_split = FALSE) {
+      new <- cursor + incr
+      if (new > bk_cols) {
+        if (can_split) {
+          new <- new %% bk_cols
+        } else {
+          # signal that we've forced a move to the next row
+          new <- -1L
+        }
+      }
+      cursor <<- new
+    }
+
+    add_start_class <- FALSE
+    for (idx in seq_len(n_kids)) {
+      if (before[idx] > 0) {
+        update_cursor(before[idx], can_split = TRUE)
+        add_start_class <- TRUE
+      }
+
+      this_width <- min(widths[idx], bk_cols)
+
+      start_at <- cursor
+      update_cursor(this_width, can_split = FALSE)
+      if (cursor < 0) {
+        add_start_class <- TRUE
+        start_at <- 0
+        cursor <- this_width
+      }
+
+      if (add_start_class) {
+        add_class(idx, sprintf("g-start-%s-%s", break_name, start_at + 1L))
+        add_start_class <- FALSE
+      }
+
+      add_class(idx, sprintf("g-col-%s-%s", break_name, this_width))
+
+      if (after[idx] > 0) {
+        update_cursor(after[idx], can_split = TRUE)
+        add_start_class <- TRUE
+        # There isn't an "end" class, so we just move the cursor forward
+      }
+
+      if (cursor == bk_cols) {
+        # Row is full, no start class needed even if after columns were added
+        cursor <- 0L
+        add_start_class <- FALSE
+      }
+    }
+  }
+
+  vapply(classes, paste, character(1), collapse = " ")
+}
+
+#' Define responsive breakpoints
+#'
+#' TODO: describe me
+#'
+#' @keywords internal
+#' @export
+breakpoints <- function(..., sm = NULL, md = NULL, lg = NULL) {
+  res <- rlang::list2(...)
+  if (!is.null(sm)) res$sm <- sm
+  if (!is.null(md)) res$md <- md
+  if (!is.null(lg)) res$lg <- lg
+
+  res <- dropNulls(res)
+  if (any_unnamed(res)) {
+    rlang::abort("All `breakpoints` values must be named")
+  }
+
+  # TODO: can we get breakpoint names from the breakpoint map to validate?
+  break_names <- intersect(c("xs", "sm", "md", "lg", "xl", "xxl"), names(res))
+  break_names <- c(break_names, setdiff(names(res), break_names))
```

Comment:
Let's add a comment about `intersect()` returning matches in order (of the 1st arg)